receiver/mongodbreceiver/extractor.go

```diff
@@ -0,0 +1,437 @@
+	if e.version.GreaterThanOrEqual(Mongo40) {
+		heldTimeUs, _ = digForIntValue(document, []string{"globalLock", "totalTime"})
+	} else {
+		for _, lockType := range []string{"W", "R", "r", "w"} {
+			waitTime, err := digForIntValue(document, []string{"locks", ".", "timeAcquiringMicros", lockType})
+			if err == nil {
+				heldTimeUs += waitTime
+			}
+		}
+	}
+
+	if heldTimeUs != 0 {
+		htMilliseconds := heldTimeUs / 1000
+		return htMilliseconds, nil
+	}
+
+	e.logger.Warn("unable to find global lock time")
+	return 0, errors.New("was unable to calculate global lock time")
+}
+
+func (e *extractor) addMetric(ts pdata.Timestamp, mb *metadata.MetricsBuilder, document bson.M, m mongoMetric, dbName string) {
+	switch m.dataPointType {
+	case integer:
+		e.addIntMetric(ts, mb, document, m, dbName)
+	case double:
+		e.addDoubleMetric(ts, mb, document, m, dbName)
+	default:
+		e.logger.Debug("unknown data point type", zap.String("name", m.name))
+	}
+}
+
+func (e *extractor) addIntMetric(ts pdata.Timestamp, mb *metadata.MetricsBuilder, document bson.M, metric mongoMetric, dbName string) {
+	value, err := digForIntValue(document, metric.path)
+	if err != nil {
+		e.logger.Debug("unable to find metric", zap.String("name", metric.name))
+		return
+	}
+
+	if metric.conversionFactor != 0 {
+		value = metric.convertToAppropriateValue(value).(int64)
+	}
+
+	switch metric.name {
+	// mongodb.collection.count
+	case collectionCount:
+		mb.RecordMongodbCollectionCountDataPoint(ts, value, dbName)
+	// mongodb.connection.count
+	case connections:
+		mb.RecordMongodbConnectionCountDataPoint(ts, value, dbName, metric.staticAttributes[metadata.A.ConnectionType])
+	// mongodb.data.size
+	case dataSize:
+		mb.RecordMongodbDataSizeDataPoint(ts, value, dbName)
+	// mongodb.index.count
+	case indexes:
+		mb.RecordMongodbIndexCountDataPoint(ts, value, dbName)
+	// mongodb.index.size
+	case indexSize:
+		mb.RecordMongodbIndexSizeDataPoint(ts, value, dbName)
+	// mongodb.memory.usage
+	case memUsage:
+		mb.RecordMongodbMemoryUsageDataPoint(ts, value, dbName, metric.staticAttributes[metadata.A.MemoryType])
+	// mongodb.extent.count
+	case numExtents:
+		mb.RecordMongodbExtentCountDataPoint(ts, value, dbName)
+	// mongodb.object.count
+	case objects:
+		mb.RecordMongodbObjectCountDataPoint(ts, value, dbName)
+	// mongodb.storage.size
+	case storageSize:
+		mb.RecordMongodbStorageSizeDataPoint(ts, value, dbName)
+	default:
+		e.logger.Warn("attempted to add an unknown metric", zap.String("metric", metric.name))
+	}
+}
+
+func (e *extractor) addDoubleMetric(_ pdata.Timestamp, _ *metadata.MetricsBuilder, document bson.M, metric mongoMetric, _ string) {
+	_, err := digForDoubleValue(document, metric.path)
+	if err != nil {
+		e.logger.Debug("unable to find metric", zap.String("name", metric.name))
+		return
+	}
+
+	switch metric.name {
+	default:
+		e.logger.Warn("attempted to add metric datapoint was collected.", zap.String("name", metric.name))
+	}
+}
```

Comment:
I don't understand how this function is presently accomplishing anything. There are no cases in the switch statement that result in adding a metric.